### ISSUE TITLE
Regression of Gradle incremental compilation

### ISSUE THREAD

**Issue by ben-manes** [google/auto#1315]
I am observing file overwrites which should have been resolved in AutoValue 1.9 with https://github.com/google/auto/issues/1075. It was [flagged](https://github.com/ben-manes/caffeine/runs/6253021167?check_suite_focus=true#step:26:2768
) by a security tool as a possible indicator of a supply chain compromise.

>Error: StepSecurity Harden Runner: Source code overwritten file: /home/runner/work/caffeine/caffeine/simulator/build/generated/sources/annotationProcessor/java/main/com/github/benmanes/caffeine/cache/simulator/report/csv/AutoValue_PlotCsv_ChartStyle.java syscall: openat by exe: /opt/hostedtoolcache/Java_Zulu_jdk/17.0.3-7/x64/bin/java

The [build scan](https://scans.gradle.com/s/yyi6onpawwdpa) confirms this as a `stableSources` input change.
<details>
<img width="1445" alt="Screen Shot 2022-05-02 at 10 50 53 AM" src="https://user-images.githubusercontent.com/378614/166298776-073619ba-7cc4-42b7-9e40-361e22dbc778.png">
</details>

I reproduced this locally using,
```
./gradlew --console plain --no-build-cache simulator:clean simulator:compileJava \
  && ./gradlew --console plain --no-build-cache simulator:compileJava
```

<details>

```
$ ./gradlew --console plain --no-build-cache simulator:clean simulator:compileJava && ./gradlew --console plain --no-build-cache simulator:compileJava
Configuration on demand is an incubating feature.
> Task :caffeine:downloadCaffeineLocal UP-TO-DATE
> Task :caffeine:compileJavaPoetJava UP-TO-DATE
> Task :caffeine:processJavaPoetResources UP-TO-DATE
> Task :caffeine:javaPoetClasses UP-TO-DATE
> Task :simulator:clean
> Task :caffeine:generateNodes UP-TO-DATE
> Task :caffeine:generateLocalCaches UP-TO-DATE
> Task :caffeine:compileJava UP-TO-DATE
> Task :caffeine:processResources NO-SOURCE
> Task :caffeine:classes UP-TO-DATE
> Task :caffeine:compileCodeGenJava UP-TO-DATE
> Task :caffeine:processCodeGenResources NO-SOURCE
> Task :caffeine:codeGenClasses UP-TO-DATE
> Task :caffeine:jar UP-TO-DATE
> Task :simulator:downloadCaffeineLocal
> Task :simulator:compileJava

BUILD SUCCESSFUL in 40s
11 actionable tasks: 3 executed, 8 up-to-date
Configuration on demand is an incubating feature.
> Task :caffeine:downloadCaffeineLocal UP-TO-DATE
> Task :simulator:downloadCaffeineLocal UP-TO-DATE
> Task :caffeine:compileJavaPoetJava UP-TO-DATE
> Task :caffeine:processJavaPoetResources UP-TO-DATE
> Task :caffeine:javaPoetClasses UP-TO-DATE
> Task :caffeine:generateNodes UP-TO-DATE
> Task :caffeine:generateLocalCaches UP-TO-DATE
> Task :caffeine:compileJava UP-TO-DATE
> Task :caffeine:processResources NO-SOURCE
> Task :caffeine:classes UP-TO-DATE
> Task :caffeine:compileCodeGenJava UP-TO-DATE
> Task :caffeine:processCodeGenResources NO-SOURCE
> Task :caffeine:codeGenClasses UP-TO-DATE
> Task :caffeine:jar UP-TO-DATE
> Task :simulator:compileJava

BUILD SUCCESSFUL in 5s
10 actionable tasks: 1 executed, 9 up-to-dat
```
</details>

To avoid flukes, I used a resolution strategy to force the dependency to the right version if an old version was transitively referenced. I also tried to remove a convenience extension, [auto-builder](https://github.com/h908714124/auto-builder), in case it broke the incremental build. The above command still showed a rebuild locally.

This is a minor annoyance but it would be nice to have a clean bill of health from security audits.

**Comment by eamonnmcmanus** [comment]
Is this an issue with AutoValue or with Gradle? We don't have much Gradle expertise on this project, unfortunately.

**Comment by ben-manes** [comment]
I don't know, the inputs changed which would make sense as newly generated classes added to the classpath. However you would think Gradle would understand why they came about as implicit inputs and not invalidate the compilation unit. Unfortunately the compilation toolchain and that low level interaction is outside of my gradle experience. Probably if we tried something similar like MapStruct would help indicate if this is a bug in the annotation processor or the toolchain.

**Comment by ben-manes** [comment]
I was able to narrow the problem down to my build's configuration, sorry for the noise.

I copied an example which had the following. The `eclipse` plugin doesn't have `sourceSets` so that referred to the top-level and then changed the java plugin's compile input, and caused the invalidation. My mistake of googling for a snippet instead of thinking through it myself.

```
eclipse {
  sourceSets.main {
    java.srcDir "${buildDir}/generated/sources/annotationProcessor/java/main"
  }
}
```
